extensions/ql-vscode/src/databases/github-repository-finder.ts

```diff
@@ -0,0 +1,139 @@
+import {
+  API as GitExtensionAPI,
+  GitExtension,
+  Repository,
+} from "../common/vscode/extension/git";
+import { extensions, Uri } from "vscode";
+import { getOnDiskWorkspaceFoldersObjects } from "../common/vscode/workspace-folders";
+import { ValueResult } from "../common/value-result";
+
+// Based on https://github.com/microsoft/sarif-vscode-extension/blob/a1740e766122c1759d9f39d580c18b82d9e0dea4/src/extension/index.activateGithubAnalyses.ts
+
+async function getGitExtensionAPI(): Promise<GitExtensionAPI | undefined> {
+  const gitExtension =
+    extensions.getExtension<GitExtension>("vscode.git")?.exports;
+  if (!gitExtension) {
+    return undefined;
+  }
+
+  const git = gitExtension.getAPI(1);
+  if (git.state === "initialized") {
+    return git;
+  }
+
+  return new Promise((resolve) => {
+    git.onDidChangeState((state) => {
+      if (state === "initialized") {
+        resolve(git);
+      }
+    });
+  });
+}
+
+async function findRepositoryForWorkspaceFolder(
+  git: GitExtensionAPI,
+  workspaceFolderUri: Uri,
+): Promise<Repository | undefined> {
+  return git.repositories.find(
+    (repo) => repo.rootUri.toString() === workspaceFolderUri.toString(),
+  );
+}
+
+async function findRemote(repository: Repository): Promise<string | undefined> {
+  // Try to retrieve the remote 5 times with a 5 second delay between each attempt.
+  // This is to account for the case where the Git extension is still initializing.
+  for (let count = 0; count < 5; count++) {
+    const remoteName = repository.state.HEAD?.upstream?.remote ?? "origin";
+    const originRemoteUrl = repository.state.remotes.find(
+      (remote) => remote.name === remoteName,
+    )?.fetchUrl;
+    if (originRemoteUrl) {
+      return originRemoteUrl;
+    }
+
+    const firstRemoteUrl = repository.state.remotes[0]?.fetchUrl;
```

Comment:
Yes, I've added a comment.

I've also changed the heuristic slightly. Previously, if you had an upstream remote on the current branch, we would first check that upstream, and then immediately select the first remote. This skips the check for `origin`. I've now added that correctly, so that when the upstream remote is not found, it will first check `origin` and then just return the first remote.